### PR TITLE
feat: add support for server relative and templated urls 

### LINKED ISSUE #494: Relative URL gives error
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.22.0

### Plugin version

4.x.x

### Node.js version

14.x

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

20.10

### Description

According to the[ OpenAPI docs](https://swagger.io/docs/specification/api-host-and-base-path/#relative-urls) it should be possible to use a relative URL in the `servers` array.
When I do this, I get the following error on startup:

```
node:internal/url:552
  throw new ERR_INVALID_URL(input);
  ^

TypeError [ERR_INVALID_URL]: Invalid URL
    at new NodeError (node:internal/errors:371:5)
    at onParseError (node:internal/url:552:9)
    at new URL (node:internal/url:628:5)
    at /opt/deploy/node_modules/fastify-swagger/lib/spec/openapi/utils.js:73:22
    at Array.forEach (<anonymous>)
    at normalizeUrl (/opt/deploy/node_modules/fastify-swagger/lib/spec/openapi/utils.js:72:11)
    at Object.swagger (/opt/deploy/node_modules/fastify-swagger/lib/spec/openapi/index.js:40:19)
    at /opt/deploy/index.js:42:11
    at Server.wrap (/opt/deploy/node_modules/fastify/lib/server.js:86:9)
    at Object.onceWrapper (node:events:509:28) {
  input: 'api/results',
  code: 'ERR_INVALID_URL'
}

```

My config looks like this

```
{
   routePrefix: '/docs',
   openapi: {
      info: {
         title: 'Results',
         description: 'C2ST Results API',
         version: package.version,
      },
      servers: [{
         url: '/api/results',
         description: 'API Server'
      }],
      consumes: ['application/json'],
      produces: ['application/json'],
      definitions: {},
      components: {
         securitySchemes: {
            openId: {
               type: 'openIdConnect',
               openIdConnectUrl: '/auth/realms/master/.well-known/openid-configuration'
            },
         },
      },
      security: [{
         openId: ['openid', 'profile'],
      }]
   },
   uiConfig: {
      docExpansion: 'list',
      deepLinking: false,
   },
   exposeRoute: true,
}
```

I am running locally - if I change the URL to `http://localhost/api/results` all works fine, but I need to be able to support on-premise installations (different hostnames). 

### Steps to Reproduce

Use a relative url in the `openapi.servers[].url` field

### Expected Behavior

A relative url should be resolved against the host serving the swagger docs.

### LINKED ISSUE #540: OpenAPI/OAS 3 Server Variable Object url ':' before 'port' variable is invalid
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.27.0

### Plugin version

4.13.1

### Node.js version

10.15

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

20.04

### Description

According to OAS 3 spec [https://swagger.io/specification/#server-object](https://swagger.io/specification/#server-object) we should be able to use server variables including `port`, however a 500 error occurs with the following stack trace;

``` 
TypeError [ERR_INVALID_URL]: Invalid URL: https://authz.localdomain:{port}/{basePath}
    at onParseError (internal/url.js:241:17)
    at new URL (internal/url.js:319:5)
    at /app/node_modules/fastify-swagger/lib/spec/openapi/utils.js:73:22
    at Array.forEach (<anonymous>)
    at normalizeUrl (/app/node_modules/fastify-swagger/lib/spec/openapi/utils.js:72:11)
    at Object.swagger (/app/node_modules/fastify-swagger/lib/spec/openapi/index.js:40:19)
    at Object.handler (/app/node_modules/fastify-swagger/lib/routes.js:101:26)
    at preHandlerCallback (/app/node_modules/fastify/lib/handleRequest.js:126:28)
    at preValidationCallback (/app/node_modules/fastify/lib/handleRequest.js:109:5)
    at handler (/app/node_modules/fastify/lib/handleRequest.js:72:7) +0ms
```

### Steps to Reproduce

Use the following server object

```
servers: [
      {
        url: 'https://authz.localdomain:{port}/{basePath}',
        description: 'Development server',
        variables: {
          port: {
            enum: ['443', '4432'],
            default: '4432',
          },
          basePath: {
            default: 'api',
          },
        },
      },
    ],
```

### Expected Behavior

No error should occur and the following url should be computed;

`https://authz.localdomain:4432/api`

### PR DESCRIPTION
**Pull request by salvdelpiano**
#### Checklist

- [x] run `npm run test` and `npm run benchmark` 
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Hello everyone, in an attempt to fix issues #494 and #540 I integrated support for `server` relative and templated urls.

When `stripBasePath` was set to true then the server url was being parsed as is using `new URL(serverUrl)`. This inevitably resulted in an error when the server url was either relative (e.g. `/foo/bar`) or templated (e.g. `http://localhost:{port}/{basePath}`). 

The changes introduced by this PR solve the issue by first resolving the server url (by replacing templated variables with the corresponing values) and then by using it as is if relative or otherwise parsing it through `new URL(serverUrl)` in order to extract the pathname.

Fixes #494, Fixes #540

### CODE REVIEW

File: lib/spec/openapi/utils.js [hunk 1]
```diff
@@ -68,18 +68,47 @@ function prepareOpenapiObject (opts) {
...
+function resolveServerUrls (servers) {
+  const resolvedUrls = []
+
+  servers = Array.isArray(servers) ? servers : []
+  for (const server of servers) {
+    const originalUrl = server.url
+    const variables = server.variables
+
+    let url = originalUrl
+    const findVariablesRegex = /{(.*?)}/g // As for OpenAPI v3 spec url variables are named in brackets, e.g. {foo}
```
Comment (Uzlopak) [inline lib/spec/openapi/utils.js:91]: I think we can move this variable outside the for loop

File: lib/spec/openapi/utils.js [hunk 2]
```diff
@@ -68,18 +68,47 @@ function prepareOpenapiObject (opts) {
...
+    const matches = url.matchAll(findVariablesRegex)
+
+    for (const match of matches) {
```
Comment (Uzlopak) [inline lib/spec/openapi/utils.js:94]: Cant we do and remove lines 95 and 96?

```suggestion
    for (const [variableNameInBrackets, variableName] of matches) {
```

File: lib/spec/openapi/utils.js [hunk 3]
```diff
@@ -68,18 +68,47 @@ function prepareOpenapiObject (opts) {
...
+      const variableNameInBrackets = match[0]
+      const variableName = match[1]
+      const variableValue = variables?.[variableName]?.default
```
Comment (Uzlopak) [inline lib/spec/openapi/utils.js:97]: can we remove the variable-prefix?

variableNameInBrackets => nameInBrackets
variableName => name
variableValue => value

**Review by Uzlopak — CHANGES_REQUESTED** [review]
just some remarks and questions

**Review by Uzlopak — APPROVED** [review]
LGTM

File: lib/spec/openapi/utils.js [hunk 4]
```diff
@@ -68,18 +68,45 @@ function prepareOpenapiObject (opts) {
...
+    const variables = server.variables
+
+    let url = originalUrl
+    const matches = url.matchAll(findVariablesRegex)
+
+    for (const [nameInBrackets, name] of matches) {
+      const value = variables?.[name]?.default
+
+      if (value === undefined) {
+        throw new Error(`Server url ${originalUrl} could not be resolved. Make sure to provide a default value for each url variable.`)
```
Comment (Fdawgs) [inline lib/spec/openapi/utils.js:98]:
```suggestion
        throw new Error(`Server URL ${originalUrl} could not be resolved. Make sure to provide a default value for each URL variable.`)
```

**Review by Fdawgs — CHANGES_REQUESTED** [review]
Does the documentation need updating to account for this feature?

**Comment by salvdelpiano** [comment]
@Fdawgs Hello there and thank you for the review.

Regarding the documentation I don't think it needs to be updated. This is more of a bug fix and the related behavior was already kind of expected but didn't work properly.

**Review by Fdawgs — APPROVED** [review]

**Comment by salvdelpiano** [comment]
@Uzlopak thank you for merging this.

I have a small request: I'm participating in hacktoberfest 2022 and it would be very cool to have this PR count toward the monthly PRs goal over there. Is there a chance you can label this PR with the `hacktoberfest-accepted` label?

**Pull request merged**